src/common/group_data.h

```diff
@@ -74,6 +84,34 @@ class ParallelGroupBuilder {
     }
     trptr[offset_key] += nelem;
   }
+
+
+
+  /*! \brief step 3: initialize the necessary storage */
```

Comment:
So we have 2 different step 3 now?

Reply:
@trivialfis thanks for review

So, for Airline-ohe data the majority of speed up comes from increased number of threads (with applied fix to remove false sharing). No significant improvements from specialization (~10%):

 DMatrix create time,s | Master  | This PR | This PR without specialized (`InitStorage`,`InitBudged`)
-- | -- | -- | --
airline-ohe contiguous numpy | 6.7 | **1.8** | **1.99**

But for Higgs data increasing number of threads (without specialized `InitStorage`,`InitBudged` method) leads to unnecessary increasing of allocated buffers sizes and time to fill and work with it.
As a result only specialization could help in this case to get declared improvements:
 DMatrix create time,s | Master  | This PR | This PR without specialized (`InitStorage`,`InitBudged`)
-- | -- | -- | --
higgs1m contiguous numpy | 0.28 | **0.08** | **0.24**

Without `InitStorage`,`InitBudged` specialization it's impossible to apply more effective memory usage as threading is implemented for abstract lines(columns or rows).

Different implementation of `InitStorage` method is required to handle optimized case (common buffer was allocated for all threads).

Reply:
@trivialfis as we have the case where specialization gives significant improvement it would be better to keep it.

Reply:
Wow, that's a factor of 4.  However, I would still want to minimize the specialization (maybe I can help).  I have been going through the code base and trying to unify the code.  A few examples:

- We have 4 different enumeration code doing similar things.  This is a huge blocker for categorical data.
- We have 4 different partitioning code.  This is a huge blocker for adaptive variate tree.
- We have 6 entries for quantile sketching.  Blocker for categorical data.
- We have 4 different histogram building code, blocker for multi target tree.
- We have 2 column sampler, blocker for feature weights.

I can go on with other features that we can have but haven't been able to do.  But as you can see if we keep specializing code XGBoost will eventually sink because failing to catch up with the machine learning algorithms (and XGBoost is all about machine learning algorithm).  That's the primary reason I haven't been able to allocate time to optimize our GPU distributed pipeline.  Because the major target is machine learning.  If someone have a better learning algorithm, or simply better heuristic, all the optimization doesn't matter because the new algorithm will beat you, 2 more percent of accuracy can overweight all the optimization (people go with pure Python library instead of optimized libraries for features, they can simply use a cloud to scale the training when needed with a bit more money).  And it would be difficult for us to catch up with the algorithm development.

For the case of DMatrix, this seems irrelevant.  But I have a strong feeling that it might eventually lead to issues.  For example, there are algorithms for determining data encoding by cross validation at early pipeline, DMatrix might be the place for it to happen (I hope it will be sketching, but I don't know yet).  So please keep in mind that, every specialization has its cost.  I made the mistake for CPU sketching, now I have to spent my time to fix my mess and your time to review it.

Reply:
Not trying to be a pain here.  The idea I want to convey is, optimization is absolutely awesome, but we should make sure that there's no other way than specialization.  For example, in general pipeline, float64 is the default precision for numpy.  `ctypes` has to allocate an array of `float32` and iterate through it to make a copy before passing it to XGBoost, which is quite expensive. (I vaguely remember for a large scipy csr_matrix it takes half of the initialization time, and double the memory usage).  Making a API accepting a general void data buffer can remove that in the pipeline.  Also, for `hist` algorithm and possibly `approx`, constructing DMatrix can be completely skipped because it uses the gradient index for training instead of accessing data itself.  So I believe there are plenty of ways to make optimization without making these small specializations.  When specialization is indeed needed, we can limit the scale.  For example the `SetIndexDataForSparse` function currently in XGBoost can be completely merged with dense version if it accept an functor for getting index with offset, the functor would be a one liner.

Reply:
@trivialfis I totally agree with you. So I reverted specification of `InitBudged/Storage` methods to simplify support of the code.
So we have next improvements results:

DMatrix create time,s | Master | Optimized | Speed up | Optimized without specification | Speed up
-- | -- | -- | -- | -- | -- 
airline-ohe contiguous numpy | 6.6 | 1.8 | 3.67 |1.99 | 3.32
msrank10k | 3.9 | 1.16 | 3.36 | 1.61 | 2.42
mortgage1Q | 4.06 | 1.1 | 3.69 | 2.93 | 1.39

Reply:
@trivialfis It's a great idea to skip DMatrix construction for `hist` method,  I think we could start to implement it.
And can we consider this option for prediction stage also? In some cases DMatrix creation take more than half of inference time (it's very important to have fast prediction in deployed application). Would it be applicable if we introduce possibility to perform inference from pandas/numpy directly without any memcopy?

Otherwise we have potential benefits from specialization and I'd propose to have `ParallelGroupBuilder` specialization for adapters, or methods provided by `ParallelGroupBuilder` class should be as part of concrete `AdapterBatch` class for example.

Reply:
> And can we consider this option for prediction stage also?

No, you can run prediction on training dataset with gradient index, but not any other dataset.  But we do have an inplace prediction function.